### ISSUE TITLE
Multiplayer server near crash

### ISSUE THREAD

**Issue by MeSoLost** [pWn3d1337/Techguns#44]
I've started using your beautiful mod on my server (MeSoLost in Space! Technic launcher) and I've found a problem with the flag entity when the world has been pre generated.  When a client comes within range of the entity the server will not crash.  The main thread just stops dead until everyone times out then it wont let anyone back in the game and the game both appears to tick (command helper events and worldboarder events still getting logged) but the main thread is just dead stuck.

I downloaded the offending regions and noticed in MCEdit that the data array which I assume is the spawning location info appears to be corrupted and/or duplicated. (normal blocks only display one very long numerical value but the "broken" flags show 2 or 3 or more of these large values)  I can confirm that this **IS** the cause of the issue (the flag, not necessarily my diagnosis) because after i erase the flags and upload the repaired region everyone can again join the server with no fatal lockup and subsequent "server took too long to tick >90000ms" auto restart.

I have at least one region file saved with a bad copy of the flag in it and I can save more if you'd like.  Also would it be possible to get some kinda specs on that array so I could potentially repair the flags instead of being forced to remove them?

**Comment by pWn3d1337** [comment]
`@Override
	public void writeEntityDataToNBT(NBTTagCompound tags) {
		super.writeEntityDataToNBT(tags);
		tags.setShort("sizeX",sizeX);
		tags.setShort("sizeZ",sizeZ);
		tags.setShort("xoffset",xoffset);
		tags.setShort("yoffset",yoffset);
		tags.setShort("zoffset",zoffset);
		tags.setIntArray("spawnPositions", spawnPositions);
		
		if(npcSpawnList !=null){
			tags.setShort("npcListSize", (short) npcSpawnList.length);
			for (int i = 0; i < npcSpawnList.length; i++) {
				tags.setString("npcName"+i, npcSpawnList[i]);
			}
		}
	}`

That's the flag's nbt write method.

sizeX and sizeZ is the size of the camp (used as area to count npcs)
the offsets are the relative position from the flag to the camp center.

SpawnPositions is an int array that has the coordinates on where to spawn the npcs. (the coordinates are x1,y1,z1, x2,y2,z2 ...)

and the npcSpawnList is the name of the npc to spawn, this is just the army soldier in camps currently.

It might be possible that a flag has damaged nbt-tags. But you can't manually place new flags, that crashes the game, the flag gets information during the worldspawn it needs. (the positions, size, etc).

**Comment by MeSoLost** [comment]
is there any way to detect the bad tags and have it attempt to auto repair (partial regenerate or single block only somehow) or have it detect the error and remove the entity? (or just post an error in the log and exclude that entity location from further processing this session)  Gonna be a pain to have to delete them every time we find one not to mention I think it's kinda lame to basicly just give the player a base to loot =\

**Comment by pWn3d1337** [comment]
Yes send me both please. But i'm not sure if I can repair it. The easy fix is to remove the flags. There is not much to loot in military bases anyway in 1.1.3 version. Military base loot got added in 1.2 alpha.

**Comment by MeSoLost** [comment]
ok guess the post didnt take =\ here's the map files.  Had to geren my map on monday, finally found it was the flags and erased 3 yesterday, and already found another one today.  looking at this one that looks closer to the code snip it appears as though the bad ones have bad spawn position array.  You're gonna need to have something in your mod that can identify a bad flag and just not even attempt to run the tile entity or have it auto repair it'self even if it only spawns solders right next to the flag after the repair.  Something odd is happening in there that not even the busted tile entity remover option in the core server config can see or else my server would just auto-delete rather than crash.  Hers the map, just under 11mb download (12mb region file) so rather than get yelled at by something I'll just dropbox it to ya.

https://www.dropbox.com/s/tx5app7sco4cc93/world.rar?dl=0

**Comment by MeSoLost** [comment]
Somehow i dont think you intended for this to happen, dunno if you fixed it yet but i have the latest version i can get from curse or forums i've found so if you'd rather have me testing on a different version for a live server hook me up with a link ^_^
![2017-03-02_20 35 27](https://cloud.githubusercontent.com/assets/20606168/23538522/af32a224-ff88-11e6-8eb1-ebc80429823f.png)

**Comment by pWn3d1337** [comment]
No matter what exactly happened, It should not happen. I will look into the flag code and see if I can improve it. For now, best option is to remove bugged flags. You can use the 1.2 alpha on curseforge if you want. It should not be less stable than the 1.1.3 version. It has a small improvement on the flags (potential crash fix)

**Comment by MeSoLost** [comment]
![screenshot 2017-03-03 16 49 20](https://cloud.githubusercontent.com/assets/20606168/23577041/17468a24-006a-11e7-85ee-b02a7677082d.png)
![screenshot 2017-03-03 17 14 00](https://cloud.githubusercontent.com/assets/20606168/23577040/17459056-006a-11e7-8330-c6d25c75a99d.png)

Images of 2 that came out bad.  Dunno if the larger picture one was going to crash but the smaller one definitely yes. Player random ported right next to it.  I'll be sure to hit curseforge instead of forum or curse sites when i do my update sweep here soon.  Absolutely FANTASTIC job by the way.  I especially love the idle turrets scanning for targets ^_^ but I havent gotten much time to play as I've been fixing bugs and updating some of my servers code to use the UN name instead of item id number values. (even old farts do stupid things at times =p )

**Comment by pWn3d1337** [comment]
Ok, the first pics looks bad. There are definitely wrong values in the spawn positions. The positions are offsets and 4million is definitely not correct. Can you show me all tags of the first pic please. I want to see if the size or something is problematic. 4294967295 is 2^32 -1 this is UNSIGNED Integer Maxvalue. Java has only signed ints so this value should not be able to be stored in an int. Maybe the flag wants so spawn NPCs at such coordinates and fucks up the server with it.

About the spaces: they are only displayed that way, an int array can't have spaces.

The second one looks fine.

Flag optimization is now officially on the TODO list for 1.2 (non-alpha) release.

**Comment by denwd** [comment]
The same problem, I ask you to speed up the solution of the problem. 40+k chunk load

**Comment by pWn3d1337** [comment]
I can't really confirm there is a problem, the screenshot with the high numbers seems to be just an incorrect display of negative numbers.

**Comment by KupchinskiyVasiliy** [comment]
CampFlagTileEnt.java

`AxisAlignedBB bb = AxisAlignedBB.getBoundingBox(posX, posY, posZ, (double) this.sizeX, 20.0D, (double) this.sizeZ).expand(8.0D, 0.0D, 8.0D);`

Change to:
`AxisAlignedBB bb = AxisAlignedBB.getBoundingBox(posX, posY, posZ, posX + (double) this.sizeX, 20.0D, posZ + (double) this.sizeZ).expand(8.0D, 0.0D, 8.0D);`

This fixes problem with BIGGEST bounding box for example:
`box[-4096.0, 6.0, -4106.0 -> 62.0, 20.0, 29.0]`
This big bounding box causes a large number of chunks load and in sum main thread stuck for a long time.

**Comment by edxp** [comment]
Is this ticket still alive? We had to disable SpawnStructures because there were just too many bases in the Nether (just above the bedrock) and some of them eventually crash the server.

**Comment by fogss** [comment]
I also have this problem
